client.go

```diff
@@ -417,24 +421,39 @@ func (c *Client) send(r *Request) (*amqp.Delivery, error) {
 	// This is where we get the responses back.
 	// If this request doesn't want a reply back (by setting Reply to false)
 	// this channel will get a nil message after publisher has Published the
-	// message.
-	r.response = make(chan *amqp.Delivery)
-	correlationID := uuid.Must(uuid.NewV4()).String()
-
-	// Set the correlation id on the publishing.
-	r.Publishing.CorrelationId = correlationID
+	// message. If the request has been used for streaming and is being reused
+	// we might already have a response channel which we don't want to
+	// override.
+	if r.response == nil {
+		r.response = make(chan *amqp.Delivery)
+	}
 
 	// This is where we get any (client) errors if they occure before we could
 	// even send the request.
 	r.errChan = make(chan error)
 
+	correlationID := uuid.Must(uuid.NewV4()).String()
+
+	c.mu.Lock()
+	// Set the correlation id on the publishing.
+	r.Publishing.CorrelationId = correlationID
+
 	// Ensure the responseConsumer will know which chan to forward the response
 	// to when the response arrives.
-	c.mu.Lock()
 	c.correlationMapping[correlationID] = r.response
 	c.mu.Unlock()
 
 	defer func() {
+		if r.stream {
+			// Store the correlation ID in the history list for the request object.
+			// This is later used to remove them from the correlationMapping when
+			// ending a streamed request.
+			r.correlationIDs = append(r.correlationIDs, correlationID)
+			r.closeFunc = generateCloseFunc(c, r)
```

Comment:
Can this be reused with non-streaming requests? Or can we avoid clearing the correlationIDs until it's full "enough"?